transports/janus_http.c

```diff
@@ -1676,103 +1647,101 @@ int janus_http_admin_handler(void *cls, struct MHD_Connection *connection, const
 	g_free(request);
 	*con_cls = NULL;   
 }
 
+static ssize_t janus_http_response_callback(void *cls, uint64_t pos, char *buf, size_t max) {
+	janus_http_msg *request = (janus_http_msg *)cls;
+	if(request == NULL || request->response == NULL)
+		return MHD_CONTENT_READER_END_WITH_ERROR;
+	if(pos >= request->resplen)
+		return MHD_CONTENT_READER_END_OF_STREAM;
+	size_t bytes = request->resplen - pos;
+	if(bytes > max)
+		bytes = max;
+	memcpy(buf, request->response + pos, bytes);
+	return bytes;
+}
+
 /* Worker to handle notifications */
-int janus_http_notifier(janus_http_msg *msg, int max_events) {
+static int janus_http_notifier(janus_http_msg *msg) {
 	if(!msg || !msg->connection)
 		return MHD_NO;
-	struct MHD_Connection *connection = msg->connection;
+	int max_events = msg->max_events;
 	if(max_events < 1)
 		max_events = 1;
 	JANUS_LOG(LOG_DBG, "... handling long poll...\n");
-	struct MHD_Response *response = NULL;
-	int ret = MHD_NO;
 	guint64 session_id = msg->session_id;
 	janus_mutex_lock(&sessions_mutex);
 	janus_http_session *session = g_hash_table_lookup(sessions, &session_id);
 	janus_mutex_unlock(&sessions_mutex);
 	if(!session || session->destroyed) {
 		JANUS_LOG(LOG_ERR, "Couldn't find any session %"SCNu64"...\n", session_id);
-		response = MHD_create_response_from_buffer(0, NULL, MHD_RESPMEM_PERSISTENT);
-		janus_http_add_cors_headers(msg, response);
-		ret = MHD_queue_response(connection, MHD_HTTP_NOT_FOUND, response);
-		MHD_destroy_response(response);
-		return ret;
+		/* TODO return error */
+		MHD_resume_connection(msg->connection);
+		return -1;
 	}
-	gint64 start = janus_get_monotonic_time();
-	gint64 end = 0;
 	json_t *event = NULL, *list = NULL;
-	gboolean found = FALSE;
-	/* We have a timeout for the long poll: 30 seconds */
-	while(end-start < 30*G_USEC_PER_SEC) {
-		if(session->destroyed)
+	while((event = g_async_queue_try_pop(session->events)) != NULL) {
+		if(session->destroyed || g_atomic_int_get(&stopping)) {
+			/* TODO return error */
+			JANUS_LOG(LOG_ERR, "Session %"SCNu64" destroyed...\n", session_id);
+			MHD_resume_connection(msg->connection);
+			return -1;
+		}
+		if(max_events == 1) {
 			break;
-		event = g_async_queue_try_pop(session->events);
-		if(session->destroyed || g_atomic_int_get(&stopping) || event != NULL) {
-			if(event == NULL)
-				break;
-			/* Gotcha! */
-			found = TRUE;
-			if(max_events == 1) {
-				break;
-			} else {
-				/* The application is willing to receive more events at the same time, anything to report? */
-				list = json_array();
-				json_array_append_new(list, event);
-				int events = 1;
-				while(events < max_events) {
-					event = g_async_queue_try_pop(session->events);
-					if(event == NULL)
-						break;
-					json_array_append_new(list, event);
-					events++;
-				}
+		} else {
+			/* The application is willing to receive more events at the same time, anything to report? */
+			list = json_array();
+			json_array_append_new(list, event);
+			int events = 1;
```

Comment:
is this a bug? 
sounds like need to move events to out of while and add events++

Reply:
Good catch, this was a copy/paste from the original code, which didn't have a while there. I'll fix it shortly.